dataset/src/test/scala/frameless/AsTests.scala

```diff
@@ -23,4 +23,27 @@ class AsTests extends TypedDatasetSuite {
     check(forAll(prop[String, Int] _))
     check(forAll(prop[Long, Int] _))
   }
+
+  test("as[X2[X2[A, B], C]") {
```

Comment:
a lot of duplication here with the previous test, but I couldn't see an easy way to not repeat myself here.